File: 프로그래머스/2/49994. 방문 길이/방문 길이.py
def solution(dirs):    
    d = {
        "U" : (1, 0),
        "L" : (0, -1),
        "R" : (0, 1),
        "D" : (-1, 0)
    }
    
    y, x = 5, 5
    visited = set()
    
    for dir in dirs:
        dy, dx = d[dir]
        ny, nx = y + dy, x + dx
        
        if 0 <= ny < 11 and 0 <= nx < 11:
            if (y, x, ny, nx) not in visited and (ny, nx, y, x) not in visited:
                visited.add((y, x, ny, nx))
            y, x = ny, nx
    
    return len(visited)

File: 프로그래머스/2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    public int solution(String dirs) {
        Map<Character, int[]> d = Map.of('U', new int[]{1,0}, 
                                    'L', new int[]{0,-1},
                                    'R', new int[]{0,1},
                                    'D', new int[]{-1,0});
        
        int y = 5;
        int x = 5;
        Set<List<Integer>> visited = new HashSet<>();
        
        for (char dir : dirs.toCharArray()) {
            int[] dydx = d.get(dir);
            int ny = dydx[0] + y;
            int nx = dydx[1] + x;
            
            if (0 <= ny && ny < 11
               && 0 <= nx && nx < 11) {
                visited.add(List.of(y, x, ny, nx));
                visited.add(List.of(ny, nx, y, x));
                y = ny;
                x = nx;
            }   
        }
        return visited.size() / 2;
    }
}